File: src/main/java/rs/kunpero/fatpak/service/UserCacheService.java
package rs.kunpero.fatpak.service;

import com.slack.api.methods.MethodsClient;
import com.slack.api.methods.SlackApiException;
import com.slack.api.methods.request.users.UsersListRequest;
import com.slack.api.methods.response.users.UsersListResponse;
import com.slack.api.model.User;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static java.util.stream.Collectors.toConcurrentMap;

@Service
@Slf4j
public class UserCacheService {
    private static Map<String, User> userCache;

    private final MethodsClient methodsClient;

    public UserCacheService(MethodsClient methodsClient) {
        this.methodsClient = methodsClient;
    }

    @Value("${slack.access.token}")
    private String accessToken;

    @PostConstruct
    public void init() throws IOException, SlackApiException {
        final List<User> members = retrieveUserList();
        userCache = members.stream()
                .filter(user -> !user.isDeleted())
                .collect(toConcurrentMap(user -> "@" + user.getName(), user -> user));
        log.info("Initialized userCache: [{}]", userCache);
    }

    public String getUserIdFromCache(final String username) throws IOException, SlackApiException {
        final User user = tryToFind(username);

        if (user == null) {
            return null;
        }
        return user.getId();
    }

    public User getUser(String username) throws IOException, SlackApiException {
        return tryToFind(username);
    }

    private User tryToFind(String username) throws IOException, SlackApiException {
        final User user = userCache.get(username);

        if (user == null || user.getId() == null) {
            log.info("Username [{}] was not found in cache. Try to find it again using slack-api", username);
            final List<User> members = retrieveUserList();
            Optional<User> optionalUser = members.stream()
                    .filter(u -> username.equals(u.getName()))
                    .findFirst();
            if (optionalUser.isPresent()) {
                userCache.put(username, optionalUser.get());
                return optionalUser.get();
            } else {
                log.warn("No username [{}] was found in the workspace", username);
                return null;
            }
        }
        return user;
    }

    private List<User> retrieveUserList() throws IOException, SlackApiException {
        UsersListRequest request = UsersListRequest.builder()
                .token(accessToken)
                .build();
        UsersListResponse response = methodsClient.usersList(request);

        return response.getMembers();
    }
}

File: src/main/kotlin/rs/kunpero/fatpak/service/UserCacheService.kt
package rs.kunpero.fatpak.service

import com.slack.api.methods.MethodsClient
import com.slack.api.methods.request.users.UsersListRequest
import com.slack.api.model.User
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.*
import java.util.stream.Collectors
import javax.annotation.PostConstruct

@Service
class UserCacheService(
    private val methodsClient: MethodsClient,
    @Value("\${slack.access.token}")
    private val accessToken: String
) {
    companion object {
        @JvmStatic
        private val log: Logger = LoggerFactory.getLogger(UserCacheService::class.java)

        @JvmStatic
        private lateinit var userCache: MutableMap<String, User>
    }

    @PostConstruct
    fun init() {
        val members: List<User> = retrieveUserList()
        userCache = members.stream()
            .filter { user: User -> !user.isDeleted }
            .collect(
                Collectors.toConcurrentMap({ user: User -> "@${user.name}" }) { user: User -> user })
        log.info("Initialized userCache: [{}]", userCache)
    }

    fun getUserIdFromCache(username: String): String? {
        val user: User = tryToFind(username) ?: return null
        return user.id
    }

    fun getUser(username: String): User? {
        return tryToFind(username)
    }

    private fun tryToFind(username: String): User? {
        val user: User? = userCache[username]
        if (user?.id.isNullOrEmpty()) {
            log.info("Username [{}] was not found in cache. Try to find it again using slack-api", username)
            val optionalUser: Optional<User> = retrieveUserList().stream()
                .filter { u: User -> (username == u.name) }
                .findFirst()
            return if (optionalUser.isPresent) {
                userCache[username] = optionalUser.get()
                optionalUser.get()
            } else {
                log.warn("No username [{}] was found in the workspace", username)
                null
            }
        }
        return user
    }

    private fun retrieveUserList(): List<User> {
        val request = UsersListRequest.builder()
            .token(accessToken)
            .build()
        return methodsClient.usersList(request).members
    }
}